### ISSUE TITLE
Unable to proxy a service

### ISSUE THREAD

**Issue by rawlingsj** [GoogleCloudPlatform/kubernetes#11296]
Seeing a strange issue when trying to proxy a service running on GKE, api version v0.21.2

`/api/v1/proxy/namespaces/default/services/myservice/index.html`
```
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {},
  "status": "Failure",
  "message": "no endpoints available for \"myservice\"",
  "code": 500
}
```
The service is definitely there and can be seen using 
`/api/v1/namespaces/default/services/myservice/`

If going directly to the pod then it works..
`/api/v1/proxy/namespaces/default/pods/myservice-w4l8b/index.html`

Is there a configuration that I'm missing on the service to allow it to be proxied or perhaps another reason that it doesn't seem to work?

**Comment by erictune** [comment]
Can you confirm that your service's selector is selecting any pods?
Run `kubectl get services`, look for the row of your service, then in the `SELECTOR` column for something like `foo=bar`.  This is the selector for your service.  Now see if that selector is selecting any pods, like `kubectl get pods -l foo=bar`.  If no results from that, then that is your problem.

**Comment by rawlingsj** [comment]
Thanks but yes that's all working correctly.  I'm able to use the service directly and via a GKE external load balancer. The issue is only when using the proxy.

**Comment by roberthbailey** [comment]
@rawlingsj If you don't mind sending me (robertbailey [at] google.com) you cluster name / project ID I can take a look and see if there's anything obviously wrong with the proxy on your master. 

Can you confirm that the proxy works for pods but not for services?

**Comment by rawlingsj** [comment]
@roberthbailey sure I'll send the details over now, but yeah the issue is the proxy works for pods but not services.

**Comment by roberthbailey** [comment]
@rawlingsj when you created your service did you give the port a name?

Try `/api/v1/proxy/namespaces/default/services/myservice:portname/index.html`

**Comment by roberthbailey** [comment]
The proxy will only find an address for your service [if the port string matches exactly](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/registry/service/rest.go#L311) and if one isn't specified in your query it [uses the empty string by default](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/util/port_split.go#L39). If you create your service with a named port, then your queries through the proxy to that service also need to include the named port.

**Comment by rawlingsj** [comment]
Ah I see.. no I created it using the CLI and didn't give a port name.  

`kubectl expose rc myservice --create-external-load-balancer=true --port=80 --target-port=9090`

I'll give that a go now.

**Comment by thockin** [comment]
Hmm, it looks like 'expose' gives unnamed ports a name of "default".  That
seems like a misfeature to me.  Care to open a bug?  Alternately we could
make teh proxy assume that no name means "first port found" which will
handle single-port services transparently.  Or both

**Comment by rawlingsj** [comment]
That's great it is working with default, many thanks both!

`/api/v1/proxy/namespaces/default/services/myservice:default/index.html`

I'll close this issue and raise one for the misfeature to be discussed.  Thanks again.

**Comment by sputnick** [comment]
Wow I spent 3 hours trying a bunch of things to get to the UI of my cluster service and it never occurred to me that a named port was screwing with the path. Would be great to mentioned this here:
https://github.com/kubernetes/kubernetes/blob/master/cluster/addons/README.md

@roberthbailey @thockin

**Comment by thockin** [comment]
Why would that go in the addons docs?

http://kubernetes.io/docs/user-guide/accessing-the-cluster/#manually-constructing-apiserver-proxy-urls

**Comment by sputnick** [comment]
@thockin in my case I was creating a cluster-service with a UI and trying to copy the way that the common kubernetes dashboard is deployed. "Kubectl cluster-info" returns the service URL without any ":portname" ending which sends the user to a dead end (from a UX perspective). So after searching and digging the cluster-addon docs and coming up with nothing I ended up here for the answer. Funny enough someone asked about the same thing on Slack the next day I managed to point him in the right direction. It's either a case of me not RTFM or linking the two might save someone else 3 hours ;)
